tsl/src/continuous_aggs/invalidation_plugin.c

```diff
@@ -0,0 +1,535 @@
+	invalidations_insert_tuple(data, relinfo, relation, as_tuple(change->data.tp.oldtuple));
+}
+
+static List *
+parse_output_parameters(List *options)
+{
+	ListCell *option;
+	List *relinfo_list = NIL;
+
+	/*
+	 * Options are pairs of relation names and the attribute to read.
+	 *
+	 * We do not check if the "key" is a hypertable since we do not want to
+	 * rely on the TimescaleDB metadata.
+	 */
+	foreach (option, options)
+	{
+		DefElem *elem = lfirst(option);
+		const char *name = elem->defname;
+#if PG16_LT
+		List *qname = stringToQualifiedNameList(name);
+#else
+		List *qname = stringToQualifiedNameList(name, NULL);
+#endif
+		TrackRelInfo *relinfo = palloc0(sizeof(TrackRelInfo));
+		relinfo->hypertable_rv = makeRangeVarFromNameList(qname);
+		relinfo->hypertable_relid = RangeVarGetRelid(relinfo->hypertable_rv, AccessShareLock, true);
+		if (!OidIsValid(relinfo->hypertable_relid))
+			ereport(ERROR,
+					errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("relation \"%s\" does not exist", name));
+		namestrcpy(&relinfo->attname, strVal(elem->arg));
+
+		relinfo->attnum = get_attnum(relinfo->hypertable_relid, NameStr(relinfo->attname));
+		if (relinfo->attnum == InvalidAttrNumber)
+			ereport(ERROR,
+					errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("attribute \"%s\" does not exist", NameStr(relinfo->attname)));
+
+		relinfo_list = lappend(relinfo_list, relinfo);
+	}
+	return relinfo_list;
+}
+
+static void
+invalidations_write_record(HypertableInvalidationCacheEntry *entry, bool last_write,
+						   InvalidationsContext *args)
+{
+	LogicalDecodingContext *ctx = args->ctx;
+	InvalidationsPluginData *data = ctx->output_plugin_private;
+	TupleDesc tupdesc = RelationGetDescr(data->logrel);
+	Datum values[_Anum_continuous_aggs_hypertable_invalidation_log_max];
+	bool isnull[_Anum_continuous_aggs_hypertable_invalidation_log_max] = { false };
+
+	TS_DEBUG_LOG("write entry: hypertable_id=%u, lowest_modified_value=%" PRId64 ", "
+				 "greatest_modified_value=%" PRId64,
+				 entry->hypertable_relid,
+				 entry->lowest_modified_value,
+				 entry->greatest_modified_value);
+
+	/* Copy over the field to the invalidation slot */
+	values[AttrNumberGetAttrOffset(
+		Anum_continuous_aggs_hypertable_invalidation_log_hypertable_id)] =
+		ObjectIdGetDatum(entry->hypertable_relid);
+	values[AttrNumberGetAttrOffset(
+		Anum_continuous_aggs_hypertable_invalidation_log_lowest_modified_value)] =
+		Int64GetDatum(entry->lowest_modified_value);
+	values[AttrNumberGetAttrOffset(
+		Anum_continuous_aggs_hypertable_invalidation_log_greatest_modified_value)] =
+		Int64GetDatum(entry->greatest_modified_value);
+
+	MinimalTuple tuple = heap_form_minimal_tuple(tupdesc, values, isnull);
+	ExecStoreMinimalTuple(tuple, args->slot, true);
+
+	OutputPluginPrepareWrite(ctx, last_write);
+	/* Since we are not writing streamed data, we do not pass in an XID */
+	logicalrep_write_insert(ctx->out,
```

Comment:
What does ‘streaming’ mean in this context, and why aren’t we writing the streamed data? I was under the impression that we were streaming, but clearly I’ve misunderstood some parts.

Reply:
Makes sense. Thanks for explaining.

Reply:
I am also confused by what this does and role of streaming. I am not very familiar with logical decoding plugins so I have no idea what streaming means in this context. Can you expand on it in the comment?

Reply:
Looking at this again, I am confused why we are calling `logicalrep_write_insert` here. This seems to be a function from logical replication, which we are not using, but internally it is just serializing the tuple you sent.

In the binary protocol, why do all of this conversion to tuple? Can't you just serialize the struct as is?

Can't you just do this:

```
pq_sendint32(out, entry->relid);
pq_sendint64(out, entry->lower_bound);
pq_sendint64(out, entry->upper_bound);
```

Then do the reverse on the other end. This will get rid of a lot of tuple conversion and need to reference a relation. I think `data->logrel` goes away and it is not needed. Nor is the tupletableslot or tupledesc.